### ISSUE TITLE
[BUG][SDW]IO transfer timed out occurred when run suspend/resume stress test.

### ISSUE THREAD

**Issue by YvonneYang2** [thesofproject/linux#1482]
**Describe the bug**
IO transfer timed out occurred when run suspend/resume with aplay stress test. Run 300 times, error info start showing at 26th/27th time, but aplay function will not be effected. 

**To Reproduce**
1. Run "aplay -Dhw:0,0 -r48000 -c2 -fs16_le /dev/zero & sudo ./stress-s3.sh 300 1 1"

**Reproduced rate**
Run 300 times, failed 26th/27th time

**Expected result**
No error occurred during stress test

**Actual result**
IO transfer timed out occurred when run suspend/resume with aplay stress test.

**dmesg**
```
[ 1372.645565] PM: suspend exit
[ 1373.091463] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[ 1373.091630] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[ 1373.591405] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[ 1373.591567] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[ 1374.044938] asix 1-6.1:1.0 enx000ec6c686d3: link up, 100Mbps, full-duplex, lpa 0x41E1
[ 1374.624624] intel-sdw sdw-master-1: IO transfer timed out
[ 1374.624641] intel-sdw sdw-master-1: SDW Slave Addr: 20025d130800
[ 1374.624647] intel-sdw sdw-master-1: SDW Slave class_id 0, part_id 1308, mfg_id 25d, unique_id 0, version 2
[ 1374.624651] intel-sdw sdw-master-1: in sdw_assign_device_num
[ 1374.624654] intel-sdw sdw-master-1: in sdw_assign_device_num
[ 1374.624659] intel-sdw sdw-master-1: Slave already registered, reusing dev_num:1
```

[s3_test_27_logger_all.log](https://github.com/thesofproject/linux/files/3844550/s3_test_27_logger_all.log)
[test_26.log](https://github.com/thesofproject/linux/files/3844551/test_26.log)
[test_27.log](https://github.com/thesofproject/linux/files/3844552/test_27.log)
[s3_test_26_logger_all.log](https://github.com/thesofproject/linux/files/3844553/s3_test_26_logger_all.log)

**Comment by YvonneYang2** [comment]
Will retest it with #1451 to check if it can fix the issue.

**Comment by YvonneYang2** [comment]
Only run suspend/resume stress test sill can reproduce the issue. Run 100 times, failed at 24th/25th time.

[s3_test_24_logger_all.log](https://github.com/thesofproject/linux/files/3845133/s3_test_24_logger_all.log)
[s3_test_25_logger_all.log](https://github.com/thesofproject/linux/files/3845135/s3_test_25_logger_all.log)
[test_24.log](https://github.com/thesofproject/linux/files/3845136/test_24.log)
[test_25.log](https://github.com/thesofproject/linux/files/3845137/test_25.log)

**Comment by YvonneYang2** [comment]
@bardliao 
Issue still can be reproduced with latest-soundwire(413de1e ) + #1451 

Run 300 times, failed at 13th/14th/28th time

**13th time**
```
[  212.378163] sof-audio-pci 0000:00:1f.3: FW Poll Status: reg=0x14001e successful
[  212.378166] sof-audio-pci 0000:00:1f.3: ipc tx: 0x60040000: GLB_STREAM_MSG: TRIG_START
[  212.378394] sof-audio-pci 0000:00:1f.3: ipc tx succeeded: 0x60040000: GLB_STREAM_MSG: TRIG_START
[  212.416820] PM: suspend exit
[  212.857612] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[  212.857710] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  213.357568] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[  213.357664] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  214.418427] intel-sdw sdw-master-1: IO transfer timed out
```
**28th time**
```
[  500.408930] PM: suspend exit
[  500.855157] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[  500.855343] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  501.354939] sof-audio-pci 0000:00:1f.3: ipc rx: 0x90020000: GLB_TRACE_MSG
[  501.355121] sof-audio-pci 0000:00:1f.3: ipc rx done: 0x90020000: GLB_TRACE_MSG
[  502.367887] intel-sdw sdw-master-1: IO transfer timed out
[  502.367906] intel-sdw sdw-master-1: SDW Slave Addr: 0
[  502.367912] intel-sdw sdw-master-1: SDW Slave class_id 0, part_id 0, mfg_id 0, unique_id 0, version 0
[  502.367917] intel-sdw sdw-master-1: Slave Entry not found
[  502.367924] intel-sdw sdw-master-1: SDW Slave Addr: 0
[  502.367929] intel-sdw sdw-master-1: SDW Slave class_id 0, part_id 0, mfg_id 0, unique_id 0, version 0
[  502.367932] intel-sdw sdw-master-1: Slave Entry not found
```

Test recipe:kernel: https://github.com/thesofproject/linux/tree/integration/soundwire-latest commit:  413de1e + #1451 
FW: https://github.com/thesofproject/sof/commits/master  commit: decd206
tplg: same path as FW branch, sof-cml-rt711-rt1308-mono-rt715.tplg
platform: CML SDW Laptop with ALC711+ ALC1308+ ALC715 

[test_13.log](https://github.com/thesofproject/linux/files/3845833/test_13.log)
[test_14.log](https://github.com/thesofproject/linux/files/3845834/test_14.log)
[test_28.log](https://github.com/thesofproject/linux/files/3845835/test_28.log)
[s3_test_13_logger_all.log](https://github.com/thesofproject/linux/files/3845844/s3_test_13_logger_all.log)
[s3_test_14_logger_all.log](https://github.com/thesofproject/linux/files/3845845/s3_test_14_logger_all.log)
[s3_test_28_logger_all.log](https://github.com/thesofproject/linux/files/3845846/s3_test_28_logger_all.log)

**Comment by plbossart** [comment]
Are those issues seen with MSI? What if MSI is disabled (along with the I2C kconfigs to avoid the I2C issues in legacy mode)

**Comment by bardliao** [comment]
The IO transfer timed out are not happened on the active device (rt711/master 0). I will try using thread on [sdw_intel_thread()](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/soundwire/intel_init.c#L180) to prevent the transfer from one master is blocked by another master's event.

**Comment by bardliao** [comment]
Using thread on sdw_intel_thread() doesn't work. After adding some trace, I found that [intel_suspend_runtime](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/soundwire/intel.c#L1467) could be triggered between mutex_lock(&bus->msg_lock); and mutex_unlock(&bus->msg_lock); in [sdw_transfer()](https://github.com/thesofproject/linux/blob/integration/soundwire-latest/drivers/soundwire/bus.c#L236), and we will disable interrupt in intel_suspend_runtime. I guess it is the root cause of this issue.

**Comment by YvonneYang2** [comment]
Issue can be reproduced with Legacy IRQ mode.

**Comment by bardliao** [comment]
@YvonneYang2 Could you retest it with https://github.com/thesofproject/linux/pull/1496?

**Comment by YvonneYang2** [comment]
Test it with #1496 with suspend/resume stress test, issue cannot be reproduced in whether MSI IRQ or Legacy IRQ mode.  
But in MSI IRQ mode, sometimes dmesg will show one line failed info as below. (Some other failed messages will track in #1472) Run 500 times, shows about 15 times. 

failed info
```
[  752.864833] intel-sdw sdw-master-3: Slave status change
[  752.864838] intel-sdw sdw-master-3: sdw_handle_slave_status: start
[  752.864841] rt715 sdw:3:25d:715:0: sdw_modify_slave_status: signaling completion for Slave 1
[  752.864895] intel-sdw sdw-master-1: Msg ignored for Slave 1
[  752.864898] intel-sdw sdw-master-1: SDW_SCP_INTMASK1 write failed:-61
[  752.864901] intel-sdw sdw-master-1: Slave 1 initialization failed: -61
[  752.864903] intel-sdw sdw-master-1: Updating Slave 1 status
[  752.864905] intel-sdw sdw-master-1: sdw_handle_slave_status: Updating Slave 1 status done
[  752.864907] intel-sdw sdw-master-1: sdw_handle_slave_status: end
```
[test_46.log](https://github.com/thesofproject/linux/files/3856510/test_46.log)
[s3_test_46_logger_all.log](https://github.com/thesofproject/linux/files/3856511/s3_test_46_logger_all.log)

Test recipe:
kernel: https://github.com/thesofproject/linux/tree/integration/soundwire-latest commit: 1bb06ec + #1496 
FW: https://github.com/thesofproject/sof/tree/cml-009-drop-stable commit: decd206
tplg: same path as FW branch, sof-cml-rt711-rt1308-mono-rt715.tplg
platform: CML SDW Laptop with ALC711+ ALC1308+ ALC715

**Comment by plbossart** [comment]
@YvonneYang2 @bardliao I don't think the error has anything to do with MSI. I annotated the sequence and it seems to me that this is a conflict between the system resume and pm_runtime. If you look at the <<< markers, you'll see we start the pm_runtime_suspend while the devices are still being enumerated.

[test_46.log](https://github.com/thesofproject/linux/files/3881664/test_46.log)

**Comment by plbossart** [comment]
can we this with the following diff:

````diff
diff --git a/drivers/soundwire/intel.c b/drivers/soundwire/intel.c
index 33e81941dc9b..8241d8aa9bba 100644
--- a/drivers/soundwire/intel.c
+++ b/drivers/soundwire/intel.c
@@ -1556,6 +1556,8 @@ static int intel_resume(struct device *dev)
                return ret;
        }
 
+       pm_runtime_mark_last_busy(dev);
+
        dev_err(dev, "%s done\n", __func__);
 
        return ret;
````

[system_resume_mark_last_busy.diff.txt](https://github.com/thesofproject/linux/files/3881687/system_resume_mark_last_busy.diff.txt)

**Comment by YvonneYang2** [comment]
Run suspend/resume stress test 300 times with https://github.com/thesofproject/linux/issues/1482#issuecomment-557718752 in MSI IRQ mode, IO transfer timed out issue cannot be reproduced.

**Comment by bardliao** [comment]
Good analyzation @plbossart . So the point is we can't go into suspend when it is under enumeration process. Will it make more sense to call pm_runtime_mark_last_busy() in sdw_program_device_num() or sdw_handle_slave_status()?

**Comment by plbossart** [comment]
@bardliao it's difficult to do any sort of pm_runtime handling at the bus level, because we don't know if the slaves will show up or not. the proposal I made was a kind of 'optimistic' one, where we push the clock and suspend later, when in practice the devices are powered. 
When the slave shows up, we will mark it as last busy in the update_status, and as a result the parent is also marked as active.
I guess we could extend a bit and make sure that the bus is marked as last busy when a device0 shows up and is enumerated, but that shouldn't really matter. that shouldn't hurt either.

I'll try and cook a better solution later today.

**Comment by plbossart** [comment]
> Run suspend/resume stress test 300 times with [#1482 (comment)](https://github.com/thesofproject/linux/issues/1482#issuecomment-557718752) in MSI IRQ mode, IO transfer timed out issue cannot be reproduced.

yay, thanks for the good news to start the week @YvonneYang2

**Comment by YvonneYang2** [comment]
Run suspend/resume stress test 300 times with MSI IRQ mode, issue cannot be reproduced. Since commit ce8487b has been merged into integration/soundwire-latest branch, close the bug. 

Test recipe:
kernel: https://github.com/thesofproject/linux/tree/integration/soundwire-latest commit: a370cb0
FW: https://github.com/thesofproject/sof/commits/master  commit: db7671a
tplg: Same with FW branch, sof-cml-rt711-rt1308-mono-rt715.tplg
platform: CML Laptop with ALC711+ ALC1308+ ALC715 in SDW mode
